attributes: values must be comparable (not documented)

Since https://github.com/grpc/grpc-go/pull/4855, attribute values must either implement `Equal(o interface{}) bool` or be comparable.

This previously wasn't the case and this restriction isn't documented as far as I can tell.

https://pkg.go.dev/google.golang.org/grpc/attributes mentions that keys must be hashable, but does not mention that values must be comparable [if they don't implement `Equal`]. That information can maybe be derived from `If Equal is not implemented, standard equality is used to determine if the two values are equal.`, but I think it could be more clear.

I would prefer that it was possible to use un-comparable standard types (e.g. using https://pkg.go.dev/github.com/google/go-cmp/cmp as the fallback comparator), but if that isn't possible, I think there should be some documentation indicating directly that values need to be comparable.

We saw this panic in services after a gRPC version bump.

### What version of gRPC are you using?

grpc@v1.43.0

### What version of Go are you using (`go version`)?

Go playground

### What operating system (Linux, Windows, …) and version?

Go playground

### What did you do?

Executed https://go.dev/play/p/TYD55tjljGj

### What did you expect to see?

True or false

### What did you see instead?

A panic

```
panic: runtime error: comparing uncomparable type map[string]string

goroutine 1 [running]:
google.golang.org/grpc/attributes.(*Attributes).Equal(0x4887e0, 0xc00009acc0)
	/tmp/gopath3674506896/pkg/mod/google.golang.org/grpc@v1.43.0/attributes/attributes.go:93 +0x194
main.main()
	/tmp/sandbox3370463311/prog.go:16 +0x207

Program exited.
```

--- a/attributes/attributes.go
+++ b/attributes/attributes.go
@@ -69,7 +69,9 @@ func (a *Attributes) Value(key interface{}) interface{} {
 // bool' is implemented for a value in the attributes, it is called to
 // determine if the value matches the one stored in the other attributes.  If
 // Equal is not implemented, standard equality is used to determine if the two
-// values are equal.
+// values are equal. Note that some types (e.g. maps) aren't comparable by
+// default, so they must be wrapped in a struct, or in an alias type, with Equal
+// defined.
 func (a *Attributes) Equal(o *Attributes) bool {
 	if a == nil && o == nil {
 		return true

